lazygal.1
.TH LAZYGAL "1" "November 2007" "lazygal" "User Commands"
.SH NAME
lazygal \- static web gallery generator

.SH SYNOPSIS
usage: lazygal [options] albumdir

.SH DESCRIPTION
This manual page explains the
.B lazygal
program. This program is a static web gallery generator written in Python.
.SH OPTIONS
.TP
\fB\-v\fR, \fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-\-quiet\fR
Don't output anything except for errors.
.TP
\fB\-\-debug\fR
Output everything that lazygal is doing.
.TP
\fB\-o\fR DEST_DIR, \fB\-\-output\-directory\fR=\fIDEST_DIR\fR
Directory where web pages, slides and thumbs will be
written (default is current directory).
.TP
\fB\-t\fR THEME, \fB\-\-theme\fR=\fITHEME\fR
Theme name (looked up in theme directory) or theme
full path.
.TP
\fB\-\-template\-vars\fR=\fITPL_VARS\fR
Common variables to load all templates with, define as <name1>=<value1>,
<name2>=<value2> .
.TP
\fB\-\-clean\-destination\fR
Clean destination directory of files that should not
be there.
.TP
\fB\-\-check\-all\-dirs\fR
Exhaustively go through all directories regardless of
source modification time (prefer the use of
.B touch
on a source directory if you want to force its processing).
.TP
\fB\-s\fR IMAGE_SIZE, \fB\-\-image\-size\fR=\fIIMAGE_SIZE\fR
Size of images, define as <name>=<x>x<y>,..., eg.
small=800x600,medium=1024x768.
.TP
\fB\-T\fR THUMBNAIL_SIZE, \fB\-\-thumbnail\-size\fR=\fITHUMBNAIL_SIZE\fR
Size of thumbnails, define as <x>x<y>, eg. 150x113.
.TP
\fB\-q\fR QUALITY, \fB\-\-quality\fR=\fIQUALITY\fR
Quality of generated JPEG images (default is 85).
.TP
\fB\-O\fR, \fB\-\-original\fR
Include original photos in output.

.SH THEMES
A theme maps to a directory that contains at least the following items :
.TP 5
<theme>/browse.thtml
The XHTML template for the theme browse page (displaying one picture).
.TP 5
<theme>/dirindex.thtml
The XHTML template for the directory index page (pictures and subgalleries
links).

.P
This must be valid XML. See <\fIhttp://genshi.edgewall.org/wiki/Documentation/xml-templates.html\fR> for syntax.

Each file which name starts with
.B SHARED_
is copied to the web album destination under the directory
.B shared
and is aimed at being a shared ressource accross your web album (CSS
stylesheet, button, etc.).

Each file which extension starts with
.B t
is processed through the genshi text template engine before being copied.
See <\fIhttp://http://genshi.edgewall.org/wiki/Documentation/text-templates.html\fR> for syntax.

.SH ALBUM METADATA
If a directory from the source album contains a file named
.B album_description
, it is processed as a source of album metadata. See \fI/usr/share/doc/lazygal/README\fR for more information.

.SH FILES
.TP 5
~/.lazygal
User configuration directory.
.TP 5
~/.lazygal/config
User configuration file. INI like file format.
.TP 5
~/.lazygal/themes
User themes directory.

.SH SEE ALSO
The full documentation for
.B lazygal
is maintained as a README file.

More information is available on program website:
<\fIhttp://sousmonlit.dyndns.org/~niol/playa/oss:lazygal\fR>.


ChangeLog
Lazygal 0.4
  * set version to 0.4
  * output tpl gen date in current local to match i18n
  * use new out parameter in genshi templates generation (genshi >=0.5)
  * add french translation
  * support for an additionnal config file in source_dir
  * do not output format not supported message twice
  * update TODO : better i18n in templates
  * add some translatable strings in the default theme
  * Initial czech translation.
  * Avoid two strings with different format chars.
  * revert previous patch and use lenient mode for variable lookup
  * i18n support for templates
  * Use defined() as it seems to be needed with genshi 0.5.0
  * remove version in manpage
  * fix typo in comment
  * add a comment about the usage of the 't' dict in templates
  * fix typo in exception raising
  * better encoding detection for album_description
  * allow xhtml tags in album_description
  * improve README a bit on template vars
  * compute utf-16 endianess from EXIF tag endianess (patch by Matthieu Castet)
  * fix typo in source dir does not exist error
  * fix unicode EXIF UserComment (thanks Matthieu Castet for the hint on endianess)
  * charset parsing is only for the UserComment EXIF field
  * strip useless space in exif strings
  * fallback to utf-8 for jpeg comments
  * better handling of decoding errors for EXIF commendings
  * handle charset decl with no text in EXIF comment
  * Readd lost zero
  * Better parsing of field
  * Better charset handling
  * Handle charset in EXIF comment
  * Handle old EXIF comment
  * Fix \0 in comment
  * add a white frame to webalbum pics which looks better
  * add an option to choose the webalbum picture color
  * TODO i18n in tpls
  * fix filename encoding issue during dest_dir cleanup checks
  * make genshi templates play better with unicode
  * TODO video and dirgalpic bg color
  * optional directory archive with original images for easy download
  * update-po script reorganization
  * i18n support for runtime messages
  * fix lazygal module * import
  * webalbum index pagination for big directories
  * remove useless Album constructor parameter
  * Another int conversion
  * Force int conversion

Lazygal 0.3.1
  * set version to 0.3.1
  * add a script that builds a changelog from the darcs history
  * better description of touch instead of --check-all-dirs in man page
  * fix thumb alt in dirindex
  * fix thumb width/height not being outputed in dirindex
  * if dep needs build, target should be built
  * rebuild webalbumpic when pics get removed from dir
  * use static var instead of instance var for configurable webalbumpic fn

Lazygal 0.3
  * set version to 0.3
  * sort subdirs for nice ordering when displayed
  * better and simplet title for index page
  * TODO abandon this item because the shell makes it really tricky to put complex stuff in tpl vars'
  * IndexPage now depends on subdir metadata
  * try to handle conffile tpl vars with non ascii chars
  * use subdir objects rather than dirnames to populate subgal links
  * make more use of _str_humanize()
  * webalbum pic should be a dep of webalbum dir instead of index page
  * skip VCS directories in album generation
  * do not fail on directory that contains no pictures
  * fix typo in tpl var name: s/zise/size/
  * do not display other size links if there is only one image size
  * Add missing template.
  * Support for new PyExiv2
  * update TODO
  * reorder Album methods a bit to have loging all in one place
  * themes mostly through CSS and new default style cmdline option
  * make the contents of the feed item a template
  * fix original link
  * allow template vars given on the cmdline to have differente encodings
  * fail if dest is in source tree
  * add webalbum picture to feed items
  * simple filedep should not be registered as output from the target
  * fix tpl file dep by normalizing DATAPATH
  * TODO Webalbumdir done another way
  * make feed use source dirs mtimes instead of generated dirs mtimes
  * share more code between LightWebalbumDir and WebalbumDir
  * maybe fix album_description generation which was broken by previous changes
  * play a bit with picture mess params to improve effect (making the thumbs larger)
  * TODO inheritence not possible between WebalbumDir et LightWebalbumDir
  * add reported bug to TODO
  * add a todo list to the project
  * shared file tpl must depend on template file
  * build some fancier album picture
  * use some extra_files list for dir cleanup
  * fix tpl shared files generation
  * walk source tree from deep to top to be able to reuse stuff processed in subdirs
  * move _add_size_qualifier() in album so that everybody gets it
  * Support for generating metadata files
  * More intelligent quotes stripping
  * use the new copy target for ImageOriginal
  * make shared files generation a make target

Lazygal 0.2
  * bump version number
  * do not complain if elementtree is not here
  * advertise RSS feed in templates now that it seems to work
  * full RSS feed generation (not only processed directories)
  * add missing metadata dependency for index page
  * quote urls in feed for special characters
  * update copyright dates in licence headers
  * Optional lens information
  * Better model name generation
  * linux kernel compile like output, which is more readable
  * fix feed generation, sorting and maxitems
  * some tidying in album metadata to bring back guessed album pictures in index page
  * depending on prev/next for browse page should occure at prepare time
  * make sure that make targets are prepared only once
  * browse page should depend on directory in case of image deletion
  * fix indentation in previous patch
  * feed should not depend on skipped directories
  * do not load exif data in unprocessed directories
  * feed generation for new or updated galleries
  * construct template dependencies with full path
  * refactor metadata as a make target and refactor WebalbumDir attrs
  * fix browse page other size links

lazygal.py
#!/usr/bin/env python
#
# Lazygal, a lazy satic web gallery generator.
# Copyright (C) 2007-2008 Alexandre Rossi <alexandre.rossi@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import sys, os, locale, gettext
from optparse import OptionParser
import genshi.core
import ConfigParser

import lazygal
from lazygal.generators import Album, SOURCEDIR_CONFIGFILE

CONFIGFILE = '~/.lazygal/config'
CONFIGDEFAULTS = {
    'theme': 'default',
    'default-style': 'default',
    'clean-destination': 'No',
    'image-size': 'small=800x600,medium=1024x768',
    'thumbnail-size': '150x113',
    'quality': '85',
    'template-vars': '',
    'thumbs-per-page': '0',
    'webalbumpic-bg': 'transparent',
}

# i18n
LOCALES_PATH = os.path.normpath(os.path.join(os.path.dirname(__file__),
                                             'build', 'mo'))
if not os.path.exists(LOCALES_PATH):
    LOCALES_PATH = '/usr/share/locale'
gettext.install('lazygal', LOCALES_PATH, unicode=1)

locale.setlocale(locale.LC_ALL, locale.getdefaultlocale())


# Read configuration file
config = ConfigParser.ConfigParser(defaults = CONFIGDEFAULTS)
# The following will hold until the config file has more than one section.
# See http://mail.python.org/pipermail/python-list/2006-March/370021.html
config.add_section('lazygal')
config.read(os.path.expanduser(CONFIGFILE))

usage = _("usage: %prog [options] albumdir")
parser = OptionParser(usage=usage)

# The help option must be changed to comply with i18n.
parser.get_option('-h').help = _("Show this help message and exit.")

parser.add_option("", "--quiet",
                  action="store_true",
                  dest="quiet", default=False,
                  help=_("Don't output anything except for errors."))
parser.add_option("", "--debug",
                  action="store_true",
                  dest="debug", default=False,
                  help=_("Output everything that lazygal is doing."))
parser.add_option("-o", "--output-directory",
                  action="store", type="string",
                  dest="dest_dir", default=".",
                  help=_("Directory where web pages, slides and thumbs will be written (default is current directory)."))
parser.add_option("-t", "--theme",
                  action="store", type="string",
                  dest="theme",
                  default=config.get('lazygal', 'theme'),
                  help=_("Theme name (looked up in theme directory) or theme full path."))
parser.add_option("", "--default-style",
                  action="store", type="string",
                  dest="default_style",
                  default=config.get('lazygal', 'default-style'),
                  help=_("Default style to apply to the theme."))
parser.add_option("", "--template-vars",
                  action="store", type="string",
                  dest="tpl_vars",
                  default=config.get('lazygal', 'template-vars'),
                  help=_("Common variables to load all templates with."))
parser.add_option("", "--clean-destination",
                  action="store_true",
                  dest="clean_dest",
                  default=config.getboolean('lazygal', 'clean-destination'),
                  help=_("Clean destination directory of files that should not be there."))
parser.add_option("-v", "--version",
                  action="store_true",
                  dest="show_version", default=False,
                  help=_("Display program version."))
parser.add_option("", "--check-all-dirs",
                  action="store_true",
                  dest="check_all_dirs", default=False,
                  help=_("Exhaustively go through all directories regardless of source modification time."))
parser.add_option("-s", "--image-size",
                  action="store", type="string",
                  dest="image_size",
                  default=config.get('lazygal', 'image-size'),
                  help=_("Size of images, define as <name>=<x>x<y>,..., eg. small=800x600,medium=1024x768."))
parser.add_option("-T", "--thumbnail-size",
                  action="store", type="string",
                  dest="thumbnail_size",
                  default=config.get('lazygal', 'thumbnail-size'),
                  help=_("Size of thumbnails, define as <x>x<y>, eg. 150x113."))
parser.add_option("-q", "--quality",
                  action="store", type="int",
                  dest="quality",
                  default=config.get('lazygal', 'quality'),
                  help=_("Quality of generated JPEG images (default is 85)."))
parser.add_option("-O", "--original",
                  action="store_true",
                  dest="original", default=False,
                  help=_("Include original photos in output."))
parser.add_option("", "--puburl",
                  action="store", type="string",
                  dest="pub_url",
                  help=_("Publication URL (only usefull for feed generation)."))
parser.add_option("-m", "--generate-metadata",
                  action="store_true",
                  dest="metadata", default=False,
                  help=_("Generate metadata description files where they don't exist."))
parser.add_option("-n", "--thumbs-per-page",
                  action="store", type="int",
                  dest="thumbs_per_page",
                  default=config.get('lazygal', 'thumbs-per-page'),
                  help=_("Maximum number of thumbs per index page. This enables index pagination (0 is unlimited)."))
parser.add_option("-z", "--make-dir-zip",
                  action="store_true",
                  dest="dirzip", default=False,
                  help=_("Make a zip archive of original pictures for each directory."))
parser.add_option("", "--webalbum-pic-bg",
                  action="store", type="string",
                  dest="webalbumpic_bg",
                  default=config.get('lazygal', 'webalbumpic-bg'),
                  help=_("Webalbum picture background color. Default is transparent, and implies the PNG format. Any other value, e.g. red, white, blue, uses JPEG."))
(options, args) = parser.parse_args()

if options.show_version:
    print _('lazygal version %s') % lazygal.__version__
    sys.exit(0)

if len(args) != 1:
    parser.print_help()
    sys.exit(_("Bad command line."))

source_dir = args[0]
if not os.path.isdir(source_dir):
    print _("Directory %s does not exist.") % source_dir
    sys.exit(1)

# Load a config file in the source_dir root
sourcedir_configfile = os.path.join(source_dir, SOURCEDIR_CONFIGFILE)
if os.path.isfile(sourcedir_configfile):
    config.read(sourcedir_configfile)

sizes = []
size_defs = options.image_size.split(',')
for single_def in size_defs:
    name, string_size = single_def.split('=')
    x, y = string_size.split('x')
    sizes.append((name, (int(x), int(y))))

x, y = options.thumbnail_size.split('x')
thumbnail = (int(x), int(y))

album = Album(source_dir, thumbnail, sizes, quality=options.quality,
              thumbs_per_page=options.thumbs_per_page,
              dirzip=options.dirzip)

if options.tpl_vars or config.has_section('template-vars'):
    tpl_vars = {}
    if config.has_section('template-vars'):
        for option in config.options('template-vars'):
            value = config.get('template-vars', option)
            value = value.decode(locale.getpreferredencoding())
            tpl_vars[option] = genshi.core.Markup(value)
    if options.tpl_vars:
        tpl_vars_defs = options.tpl_vars.split(',')
        for single_def in tpl_vars_defs:
            name, value = single_def.split('=')
            value = value.decode(sys.stdin.encoding)
            tpl_vars[name] = genshi.core.Markup(value)
    album.set_tpl_vars(tpl_vars)

album.set_theme(options.theme, options.default_style)
album.set_original(options.original)
album.set_webalbumpic(bg=options.webalbumpic_bg)

log_level = None
if options.quiet:
    log_level = 'error'
if options.debug:
    log_level = 'debug'
if log_level != None:
    album.set_logging(log_level)

if options.metadata:
    album.generate_default_medatada()
else:
    album.generate(options.dest_dir, options.pub_url,
                   options.check_all_dirs, options.clean_dest)


# vim: ts=4 sw=4 expandtab
